### ISSUE TITLE
Draft Multi-VM Arquitecture

### ISSUE THREAD

**Issue by deeplow** [freedomofpress/dangerzone#530]
Draft up the approach of multi-VM deployment for Dangerzone in Qubes. This will not include any actual implementation and whatever solution we have in the end should accommodate two use cases:
  - usage as apart of the [SecureDrop Workstation](https://github.com/freedomofpress/securedrop-workstation/)
  - installation within a Qubes system (without SecureDrop)

I made a post on the Qubes forum summarizes the possible approaches: [
What’s the Future Of Multi-VM Applications?](https://forum.qubes-os.org/t/whats-the-future-of-multi-vm-applications-securedrop-workstation-inspired/18649/6)

**Comment by rocodes** [comment]
Just going to write a comment with some of the discussion points we had at the SD coworking week a couple weeks ago, in case they are helpful for your thinking on this.

We discussed a few themes, particularly: our priorities (eg security vs usability tradeoffs, or system properties we wanted to prioritize); and our assumptions (mostly expected use conditions). This was so that if we came to a technical decision, we could fall back on coherent/cohesive principles.

Our general direction I would say could be summarized like this (and it's very similar to conversations many folks have had already, so hat tip to @deeplow @eaon and others for thinking this through).
- Have clear boundaries between "VM management" (external: setup, config, rpc policy creation/mgmt) vs "VM configuration" (internal, specific files, scripts, etc).
  - In particular, things needed inside vms belong in packages (And Salt/dom0 orchestration/copying individual files and scripts into place in VMs via salt formulas is sort of smell/anti-pattern). 
  - This principle almost maps to "anything the adminAPI can do is external, anything else is internal," except I'm not sure that the admin API can trigger VM updates, so it's not a perfect map but it's the right idea.
- A 'good' pattern: vm mgmt/orchestration/policy management (from more privilieged domain) -> install packages + write a read-only config to a VM that is read at start time (via qubesdb) -> start a service configured per-vm (via systemd).
  - This works _very_ nicely with disposable VMs
- Inter-VM communication: we still haven't discussed in detail any kind of standardizing/best practices around inter-vm communication, and I think this is where we have the most work to do. 
  - Relevant to DZ
  - Relevant to many parts of SDW (proxy <-> tor and proxy <-> client;  eventual client <-> dz; client <-> export)
  - There are lots of open questions. (Use socket-based rpc services? use executables? Best practices for what is a service arg, what to write to stdout/err, how to pass the information we want back and forth?)
  - Particularly if we are going to have multiple different ways to export files from SDW (eg: export to USB, "export" to dangerzone, export + print, export to whatever other tip management system) it would be nice if we had a semi-standard inter-vm communication format. And then any apps or VMs that want to become part of the SD/file sanitization ecosystem would need to implement that format.

**Comment by deeplow** [comment]
## Managed Dangerzone Install


| **Short Term** (actual proposal) | **Longer Term** (let's not get blocked on these 🙏) |
|---------------------------------|----------------------------------------------------------------------------|
| Dangerzone   | Integration with [SecureDrop Workstation](https://github.com/freedomofpress/securedrop-workstation), if well implemented DZ doesn't need to be SecureDrop-aware |
| Ansible in Managment qube | — |
| Propose Policy and Mgmt qube installer on [QubesOS-Contrib](https://github.com/QubesOS-Contrib) | [Design and create salt formulas distribution method](https://github.com/QubesOS/qubes-issues/issues/1939) (but with Ansible, I guess)
| Updates via Qubes Updater <img width="20" height="22" alt="Image" src="https://github.com/user-attachments/assets/ff9a881a-f221-4db4-814b-da8dc6158e15" /> | Qubes updater has a way to detect multiVM configuration updates |

Security impact:
- (+) Ability to issue actual multiVM security updates (to avoid [manual advisories](https://github.com/freedomofpress/dangerzone/blob/main/docs/advisories/2023-10-25.md))
- (-) Custom templates on networkless qubes or in no-often-run-qubes don't have changes of flagging to dom0 that they have updates. After 7 day by default the qubes updater will suggest that the user updates these. But that's a big lag. At least this is for the conversion component of Dangerzone.

## Overview

<img width="995" height="439" alt="Image" src="https://github.com/user-attachments/assets/5cd91d34-bb4c-491d-ba04-a58c079c3208" />

### Management Qube (`dz-admin`?)
*What creates and manages the other qubes that compose Dangerzone*

- **OS:** debian-<latest>-minimal (for maintenance) 
- **VM type:** Standalone

(*) this is the most brittle part but Qubes doesn't yet have a way to manage this life-cycle. Upgrades when the template is reaching end-of-life (EOL) have to be performed inplace (thus the need for a standalone qube) 

### RPC Policy

Assuming a variation of the [proposed qrexec policy](https://github.com/QubesOS/qubes-issues/issues/10015) syntax makes it possible to have the management qube manage a policy file scoped to dz.Convert policy from any other qube.

TODO add practical steps once that is implemented.

## Life-cycle management

### First Install
User install `dangerzone-installer` via QubesOS Contrib. Under the hood it does the following...

Dom0 stage:
1. Install latest `debian-minimal` template 
2. Create standalone `dz-admin` (from `debian-minimal`)
3. Set up management RPC policies (i.e. in `/etc/qubes/policy.d/`) allowing `dz-admin` to manage it's own qubes
3. Start `dz-admin` and via `qvm-run`:
  - install dangerzone repo (hardcoded GPG key)
  - install `dangerzone-admin` package (during this step it actually does the MultiVM installation)

MultiVM installation (in `dz-admin`) ansible playbook to setup remaining qubes instructed [in INSTALL.md](https://github.com/freedomofpress/dangerzone/blob/main/INSTALL.md#qubes-os)

### MultiVM Updates

A newer version of the `dangerzone-admin` package could trigger a multiVM application update (essentially the same to the "first install"). Just run the same in postinst. 

> ⚠️ **Warning**
> This has usability implications in systems with low RAM. Because the updater is unaware that it's updating a multiVM application when it is updating `dz-admin`, the system may run out of RAM (think: installing all updates + management qube starts other qubes). This is a limitation that IMO will go away once the updater is able to recognize multiVM application templates. 

## Client Application

You'll notice that there hasn't been any mention yet of where the client application will be. This is because the client part would be installed in any template the user has wants to run Dangerzone from.

**Comment by deeplow** [comment]
Just added an image to make the point clearer.

And I have dropped the idea that the mangement VM would be an app qube based on a template. The reason why I included it originally was to make upgrades easier, but I now realized that there would be issues with doing MultiVM updates. For example, imagine that the dangerzone-admin package was to be updated (thus indicating a multiVM change in need of applying). In this case the template would need to somehow start the `dz-admin` qube (which is based on it) so it would apply the latest changes. However, the template needs to shut down for `dz-admin` to have those latest changes. Overall it would make things unpractical.

Until we have dom0's Qubes Updater taking care of calling the management qubes directly, I don't think we'll be able to have anything other than standalones. In the future, I'd like to even have disposables with on-boot configured repos and packages. But that's for another time.

**Comment by apyrgio** [comment]
Thanks a lot for the nice explainer, and for the visuals. I have a couple of comments to make sure I get the idea:

> install dangerzone-admin package (during this step it actually does the MultiVM installation)

Like a post-install hook, right?

> Set up management RPC policies (i.e. in /etc/qubes/policy.d/) allowing dz-admin to manage **it's own qubes**

In that case, who is responsible for installing `dangerzone-client` on the default template, which I assume is not managed by the management VM?

---

And one more thing I was thinking about is what's the uninstallation story in this case?

**Comment by deeplow** [comment]
> Like a post-install hook, right?

Yes, that's what I was thinking. Maybe later down the line Qubes could have some other bootstrapping process.

> > Set up management RPC policies (i.e. in /etc/qubes/policy.d/) allowing dz-admin to manage **it's own qubes**
> 
> In that case, who is responsible for installing `dangerzone-client` on the default template, which I assume is not managed by the management VM?

In this original proposal, the idea is that the user would have to install it there. Maybe one way to get around that could be to have an additional `dz-client` qube, which would be targettable by `qubes.OpenInVM` and that would be the one showing the frontend. And then another one would do the conversion itself. It's a bit wasteful of resources, though.

But I haven't given this too much though. There are probably better solutions.

> And one more thing I was thinking about is what's the uninstallation story in this case?

There are two parts:
1. Removing the `dz-admin`-managed qubes
2. Removing the management qube and RPC policies

The easy part is `1.`. The managment qube would have a playbook to remove all the qubes under its management. Since there is no state to preserve, it should be ok for full removal without much hassle.

Point `2.` could be done via RPM packaging upon removal of the dom0 boostrapping package. But in the future I'd like for there to be a Qubes-native way of doing this.

**Comment by deeplow** [comment]
Now that I think about this, until there is [upstream support for downloading and installing multiVM applications](https://github.com/QubesOS/qubes-issues/issues/1939), there could be an unofficial "sudo" management qube, with internet access. It would need manual installation via a dom0 package, but from there it could fetch more multiVM applications and install them with the correct permissions without any need for any more dom0-distributed software.

In fact, the Qubes Admin API even allows for the installation of Templates from a management qube, which would be excellent for bootstrapping other management VMs.

**Comment by almet** [comment]
Thanks for the explanation and the graphs.

My understanding of the procedure, and the packaging for this story, would be to have three different parts:

- The **`dangerzone-admin` package,** that is installed on the Management VM, which basically does the provisioning for us with ansible/salt playbooks, as install / remove scripts (instructions on how to setup the management VM, configure the disposables and have the RPC policies in place) ;
- The **`dangerzone-client` package**, which is installed in the user or app qube, which is just the GUI.
- The **`dangerzone-server` package**, which does the docs-to-pixel conversion

Did I get this right?

I don't really understand what the issue is about having a standalone for the management VM. You wrote:

> Upgrades when the template is reaching end-of-life (EOL) have to be performed inplace (thus the need for a standalone qube)

I think I don't understand what you mean by having the updates performed "inplace"?
Also, if not using standalones, we would be relying on templates (for which my understanding is that they would get the updates for us)?

**Comment by deeplow** [comment]
> My understanding of the procedure, and the packaging for this story, would be to have three different parts:
> 
> - The **`dangerzone-admin` package,** that is installed on the Management VM, which basically does the provisioning for us with ansible/salt playbooks, as install / remove scripts (instructions on how to setup the management VM, configure the disposables and have the RPC policies in place) ;
> - The **`dangerzone-client` package**, which is installed in the user or app qube, which is just the GUI.
> - The **`dangerzone-server` package**, which does the docs-to-pixel conversion
> 
> Did I get this right?

Pretty much. Although `dangerzone-server` and `dangerzone-client` can remain just one dangerzone package, as is the case today. It's just that one part runs the client, the other runs the server. But there's a extra package that's needed currently:

- The **`dangerzone-installer` package**, build for dom0, if possible, distributed [via the Qubes-contrib repo](https://www.qubes-os.org/doc/package-contributions/). This is a minimal package and all that it does is setup the dom0 policy files and create the management qube. The way I see it, this is something that would be replaced by something better once Qubes has a native way of installing multiVM applications. 

> I don't really understand what the issue is about having a standalone for the management VM. You wrote:
> 
> Upgrades when the template is reaching end-of-life (EOL) have to be performed inplace (thus the need for a standalone qube)
> 
> I think I don't understand what you mean by having the updates performed "inplace"?

See [here](https://www.qubes-os.org/doc/templates/debian/in-place-upgrade/). The gist of it is that if the `dangerzone-mgmt` is reaching its end-of-life (e.g. Debian 12), then instead of doing a regular update, it would somehow trigger a system upgrade so it could migrate to Debian 13.

The alternative is using [deferred template changes](https://github.com/qubesos/qubes-issues/issues/8070). I had put this in my original proposal above, but I later changed it because I realized a limitation with it, but now I can't recall what it was. But the idea here is that the management qube running the almost-expired Debian 12 would download set its own template to Debian 13. The next time it would start, it would be running Debian 13.
